dask/dataframe/io/parquet/arrow.py

```diff
@@ -0,0 +1,450 @@
+import json
+
+import pandas as pd
+import pyarrow as pa
+import pyarrow.parquet as pq
+from ....bytes.core import get_pyarrow_filesystem
+from ....utils import natural_sort_key, getargspec
+from ..utils import _get_pyarrow_dtypes, _meta_from_dtypes
+from ...utils import clear_known_categories
+
+from .utils import _parse_pandas_metadata, _normalize_index_columns, Engine
+
+
+def _get_md_row_groups(pieces):
+    """ Read file-footer metadata from each individual piece.
+
+    Since this operation can be p[ainfully slow in some cases, abort
+    if any metadata or statistics are missing
+    """
+    row_groups = []
+    for piece in pieces:
+        for rg in range(piece.get_metadata().num_row_groups):
+            row_group = piece.get_metadata().row_group(rg)
+            for c in range(row_group.num_columns):
+                if not row_group.column(c).statistics:
+                    return []
+            row_groups.append(row_group)
+    return row_groups
+
+
+class ArrowEngine(Engine):
+    @staticmethod
+    def read_partition(fs, piece, columns, index, **kwargs):
```

Comment:
I wonder if we perhaps don't want to rewrite the docstrings here and instead rely on the docstrings of the `Engine` superclass. 

cc @jcrist knows the right way to do this I think

Reply:
Python methods don't do this automatically, but it's easy to make a decorator to do this (or use `functools.wraps`, but we only want to copy the docstring). In skein I use this:

https://github.com/jcrist/skein/blob/3baa06cddd2b7ddc743d02abcd56217522fe8959/skein/utils.py#L218-L222

e.g.

```python
class Foo(object):
    def bar(self, a):
        """I am a base method"""
        raise NotImplementedError

class Bar(Foo):
    @implements(Foo.bar)
    def bar(self, a)
        return a + 1
```

which is similar to what pyarrow uses for their filesystem implementations

Reply:
Ah ok, I thought that there might be some abstact-base-class magic to do this class-wide.

Reply:
It looks like Python does this automatically now

```python
In [1]: class A:
   ...:     def f(self):
   ...:         """ Foo """
   ...:

In [2]: class B(A):
   ...:     def f(self):
   ...:         pass
   ...:

In [3]: B.f?
Signature: B.f(self)
Docstring: Foo
File:      ~/<ipython-input-2-482a2043862c>
Type:      function
```

https://softwareengineering.stackexchange.com/a/394672